### DISCUSSION TITLE
Создание отчётной таблицы на основе данных из связанных таблиц

### DISCUSSION THREAD

**Discussion by gorbarov** [totumonline/totum-ru-issues-and-discussions#199] [Как это сделать?] [answered]
Есть у меня табличка 1

![9e7520826c](https://user-images.githubusercontent.com/55755565/187024395-0149ed52-6952-45bb-8283-9b3ab079e1b0.png)

Она связанна с таблицами 3*, которые отмечены стрелочками. В них хранятся дополнительные сведения, необходимые в итоговом вычислении.

Мне из неё надо собрать существенно более страшную таблицу 2, которая выглядит примерно вот так, только ещё по годам 

![d3a9fd9683](https://user-images.githubusercontent.com/55755565/187024401-3b2496b7-cdf5-4655-ab44-35c5bc361c96.png)

Как это в логике тотума реализуется?

Если бы я прогал, я бы просто сджойнил таблицы нужные, а в коде уже данные по условиям разложил и обработал.
Вопрос в том, что данные в таблице 2 раскладываются в зависимости от полей, которые лежат в таблицах 3*.

Как это сделать с учётом ресурсоёмкости, таблица 1 заполняется примерно 5к строк за полтора года?

Это же тривиальная задачка для Тотума должна быть, просто я не понимаю как правильно это сделать.

У меня есть идея, но я хочу понять правильно ли я мыслю с точки зрения оптимальной скорости работы Тотума.
Мысль такая: для реализации этого нужно стащить все данные в таблицу 1 в скрытые поля, например, чтобы выбирать уже из этой таблицы за один раз?
Но всё равно не понимаю, как в цикле перебирать строки и расставлять значения в зависимости от разных условий.

Например, если в таблице 1 в строке указан проект, вид работ почасовка и должность определённая — это нужно в таблице отчётной считать прямыми трудозатратами.

Направь плиз в верную сторону.
Я пока думаю так: надо собрать данные из таблиц 3* в таблицу 1 в скрытое поле.
В таблице 2 собирать массив в хедере в какой-нибудь insert, например.

Но тут второй вопрос, насколько правильно довольно объёмную отчётную таблицу делать временной. Данные сюда пока будут раз в месяц вносится.

Но когда мы полностью на тотум переедем, то они будут уже раз в неделю раскидываться автоматом.

Заранее спасибо, хорошего дня.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Логика в Тотум в основном описывается фразой «по ведущему полю». В показанном случае ведущее поле — **Проект.**

По годам — значит это Циклы, один цикл — один год. Про это есть в мастерклассе по архитектуре: https://ru.docs.totum.online/workshops

В таблице «Табличка 1» есть привязка по году и проекту. Делаем выборку проектов по году, делаем уникальным — вот список проектов в ведущее поле.

Дальше по колонкам считаем данные. Надо посчитать сумму ФОТ, ок:

```
=: listSum(list: $list)
list: selectList(table: 'table_1'; field: 'sum'; where: 'year' = #h_year; where: 'project' = #project)
```

Если расчетов в отчете становится слишком дофига и не нужно их пересчитывать каждый раз при открытии — то коды ставятся исполняемыми при добавлении и пересчитываются по крону (ночью) с указанием `field` в функции `recalculate`.

Если есть какие-то общие для всех проектов расчеты, от которых берутся доли для конкретного проекта — общие считаются один раз в хедере и одли рядом. В строках же обращение в этом случае к хедеру.
